### PR TITLE
Create IndexVersionAllocationDecider

### PR DESCRIPTION
**Pull request by thecoop**
Create IndexVersionAllocationDecider alongside NodeVersionAllocationDecider. This uses IndexVersion to ensure we don't allocate to nodes that have different IndexVersions (but might have the same Version). Eventually, NodeVersionAllocationDecider will go away, but we're not there yet

### CODE REVIEW

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
This allocation decider exists because we need to be able to copy data (both segments and operations) from the primary to the replica, which means that the replica must understand at least everything that the primary understands. However @rjernst just mentioned to me that `IndexVersion` doesn't cover certain changes that would prevent us from copying data from a newer version to an older version, such as the introduction of new mapping types. I think we need to account for all sorts of compatibility issues in this decider.

**Comment by thecoop** [comment]
I've modified this to add `IndexVersionAllocationDecider` alongside `NodeVersionAllocationDecider`. This will improve the behaviour on serverless, but as discussed we need more work before we can replace `NodeVersionAllocationDecider` completely.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexVersionAllocationDecider.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+ */
+public class IndexVersionAllocationDecider extends AllocationDecider {
+
+    public static final String NAME = "index_version";
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        if (shardRouting.primary()) {
+            if (shardRouting.currentNodeId() == null) {
+                if (shardRouting.recoverySource() != null && shardRouting.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
```
[the remark this answers is not in the document]
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexVersionAllocationDecider.java:33]: This is copied from `NodeVersionAllocationDecider` - I'll modify it there too

**Comment by thecoop** [comment]
@DaveCTurner Could we get this in alongside NodeVersionAllocationDecider? It will prevent allocation problems when IndexVersion is bumped on serverless. We can remove VersionAllocationDecider later when we find a solution to the mapping type/option allocation problem

**Comment by thecoop** [comment]
@elasticmachine  update branch

**Comment by thecoop** [comment]
@DaveCTurner I've updated this to use the mapping update index version added by https://github.com/elastic/elasticsearch/pull/106748.

Do we also need to consider the highest index version used to write a shard, as added by https://github.com/elastic/elasticsearch/pull/106779? Or if the mappings aren't updated, does it matter? If a new shard is created as part of a write, will have the index version of the node that created it?

**Comment by thecoop** [comment]
So, `IndexVersionAllocationDeciderTests.testRollingRestart` is failing, however I'm not sure this test is valid as-is. This test is ensuring that indicies do not get moved from a higher version to a lower version node during a restart...but we're not longer considering the node versions we're moving to/from, we're now looking at the index metadata and what is actually stored.

Maybe combined with https://github.com/elastic/elasticsearch/pull/106779, might it actually be valid to move a shard from a later version node to an earlier node, if nothing has actually changed about the shard?

I've also removed the distinction between primary and replica shard behaviour, now its just considering the shard as-is.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexVersionAllocationDecider.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        if (shardRouting.primary()) {
+            if (shardRouting.currentNodeId() == null) {
+                if (shardRouting.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
+                    // restoring from a snapshot - check that the node can handle the version
+                    return isSnapshotIndexVersionCompatible((SnapshotRecoverySource) shardRouting.recoverySource(), node, allocation);
+                } else {
+                    // existing or fresh primary on the node
+                    return allocation.decision(Decision.YES, NAME, "the primary shard is new or already existed on the node");
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexVersionAllocationDecider.java:39]: This is not really true in stateless, where there is no "already existed on the node". I wonder if we need to record the use of an `IndexVersion` in metadata when allocating a shard, but suspect there could be other strategies too.

I think this is umodified by the PR and hence we could choose to ignore this for now, depending on the purpose of this PR.

I think we should at least check for the mapping updated version (which might be more of an edge case though).

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexVersionAllocationDecider.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                }
+            } else {
+                return isShardIndexVersionCompatible(shardRouting, node, allocation);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexVersionAllocationDecider.java:42]: Should this use `isIndexVersionCompatibleRelocatePrimary` as well?

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexVersionAllocationDecider.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            }
+        } else {
+            // replica - just check the index mappings version & segment updated version
+            return isShardIndexVersionCompatible(shardRouting, node, allocation);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexVersionAllocationDecider.java:46]: I wonder if this should check for same or newer index version of the primary vs replica nodes in addition to the check against metadata?

**Review by henningandersen — COMMENTED** [review]
Left a few initial comments to the decider.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexVersionAllocationDecider.java (hunk 2 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexVersionAllocationDecider.java:39]: This PR is to replace `NodeVersionAllocationDecider` with equivalent functionality, but not referencing the node version. Extensions and enhancements for stateless would be separate work

**Comment by thecoop** [comment]
I think we also need a data version check - make sure we don't allocate shards to an index version that is less than the highest index version used to write data. This needs to be stored in index metadata, which means we need some kind of listener to be notified of new commits being added, and which index version it is, and stores that version in the index metadata.

**Comment by thecoop** [comment]
As discussed, I have re-created IndexVersionAllocationDecider by copying NodeVersionAllocationDecider, replacing references to node version with maximum index version

**Review by henningandersen — COMMENTED** [review]
Looks good. Have a few comments, primarily on the test.

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/IndexVersionAllocationDeciderTests.java [hunk 5]
```diff
@@ -0,0 +1,703 @@
...
+        clusterState = startInitializingShardsAndReroute(strategy, clusterState);
+
+        for (int i = 0; i < clusterState.routingTable().index("test").size(); i++) {
+            assertThat(clusterState.routingTable().index("test").shard(i).size(), equalTo(3));
+            assertThat(clusterState.routingTable().index("test").shard(i).primaryShard().state(), equalTo(STARTED));
+            assertThat(clusterState.routingTable().index("test").shard(i).replicaShardsWithState(STARTED).size(), equalTo(2));
+        }
+    }
+
+    public void testRandom() {
```
[the remark this answers is not in the document]
Comment (thecoop) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/IndexVersionAllocationDeciderTests.java:177]: I've changed it so the IndexVersion one modifies IndexVersion, and NodeVersion modifies Version

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/IndexVersionAllocationDeciderTests.java [hunk 6]
```diff
@@ -0,0 +1,703 @@
...
+            assertThat(clusterState.routingTable().index("test").shard(i).shard(1).state(), equalTo(UNASSIGNED));
+            assertThat(clusterState.routingTable().index("test").shard(i).shard(2).state(), equalTo(UNASSIGNED));
+            assertThat(clusterState.routingTable().index("test").shard(i).shard(0).currentNodeId(), nullValue());
+            assertThat(clusterState.routingTable().index("test").shard(i).shard(1).currentNodeId(), nullValue());
+            assertThat(clusterState.routingTable().index("test").shard(i).shard(2).currentNodeId(), nullValue());
+        }
+        clusterState = ClusterState.builder(clusterState)
+            .nodes(
+                DiscoveryNodes.builder()
+                    .add(newNode("old0", VersionUtils.getPreviousVersion(), IndexVersionUtils.getPreviousVersion()))
```
[the remark this answers is not in the document]
Comment (thecoop) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/IndexVersionAllocationDeciderTests.java:273]: I've generally ensured the index version/node version tests keep the other version id the same

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/IndexVersionAllocationDeciderTests.java [hunk 7]
```diff
@@ -0,0 +1,703 @@
...
+        ShardId shard1 = new ShardId("test1", "_na_", 0);
+        ShardId shard2 = new ShardId("test2", "_na_", 0);
+        final DiscoveryNode newNode = DiscoveryNodeUtils.builder("newNode").roles(MASTER_DATA_ROLES).build();
+        final DiscoveryNode oldNode1 = DiscoveryNodeUtils.builder("oldNode1")
+            .roles(MASTER_DATA_ROLES)
+            .version(VersionUtils.getPreviousVersion(), IndexVersions.MINIMUM_COMPATIBLE, IndexVersionUtils.getPreviousVersion())
+            .build();
+        final DiscoveryNode oldNode2 = DiscoveryNodeUtils.builder("oldNode2")
+            .roles(MASTER_DATA_ROLES)
+            .version(VersionUtils.getPreviousVersion(), IndexVersions.MINIMUM_COMPATIBLE, IndexVersionUtils.getPreviousVersion())
```
[the remark this answers is not in the document]
Comment (thecoop) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/IndexVersionAllocationDeciderTests.java:328]: I've updated all of these

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/IndexVersionAllocationDeciderTests.java [hunk 8]
```diff
@@ -0,0 +1,703 @@
...
+            assertThat(clusterState.routingTable().index("test").shard(i).size(), equalTo(3));
+            assertThat(clusterState.routingTable().index("test").shard(i).primaryShard().state(), equalTo(STARTED));
+            assertThat(clusterState.routingTable().index("test").shard(i).replicaShardsWithState(STARTED).size(), equalTo(1));
+            assertThat(clusterState.routingTable().index("test").shard(i).replicaShardsWithState(UNASSIGNED).size(), equalTo(1));
+        }
+
+        clusterState = ClusterState.builder(clusterState)
+            .nodes(
+                DiscoveryNodes.builder(clusterState.nodes())
+                    .add(newNode("node3", VersionUtils.getPreviousVersion(), IndexVersionUtils.getPreviousVersion()))
```
[the remark this answers is not in the document]
Comment (thecoop) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/IndexVersionAllocationDeciderTests.java]: Updated. The one exception is deciding on snapshot allocations - which can only use index version. `NodeVersionAllocationDecider` references IndexVersion in `isVersionCompatible` - is it best to just remove that path so NodeVersionAllocationDecider always returns YES for snapshots, and leaves it up to IndexVersionAllocationDecider?

[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/IndexVersionAllocationDeciderTests.java]: I'd leave that as is for this PR. To some extent, `NodeVersionAllocationDecider` ought to use the node-version until we are certain that the index-version will work, but that is umodified by this.

**Review by henningandersen — APPROVED** [review]
LGTM.

**Pull request merged**